Website: Updates `Hds::Table` component documentation (#1198)
hashicorp/design-system: website/docs/components/table/partials/code/how-to-use.md

## How to use this component

> ### Non-sortable table with no model defined, caption defined

Reviewer: I am not sure we should qualify normal tables as "non-sortable". They're just tables.
Plus, I would remove the "caption defined" from the title, for simplicity (if you want to stress the importance of adding a caption argument, I would add a specific example in the `More examples" or have a dedicated section about it in the "Accessibility" panel ("Best practices").

```suggestion
### Table with no model defined
```

</Hds::Table>
```

#### Custom sort callback

> To implement a custom sort callback on a columns, (1) add a custom function as the value for `sortingFunction` in the column hash, and (2) include a custom `onSort` action in your table invocation. This is useful for cases where the key might not be A-Z or 0-9 sortable by default, i.e., status, and you’re otherwise unable to influence the shape of the data in the model. Here is an example, code truncated for clarity:

Reviewer: ```suggestion
To implement a custom sort callback on a columns, (1) add a custom function as the value for `sortingFunction` in the column hash, and (2) include a custom `onSort` action in your table invocation to track the sorting order and use it in the custom sorting function. This is useful for cases where the key might not be A-Z or 0-9 sortable by default, i.e., status, and you’re otherwise unable to influence the shape of the data in the model. Here is an example, code truncated for clarity:
```

Reviewer: Also, should we use a `<ol>` element here, instead of `(1)/(2)`?

```handlebars{data-execute=false}
<!-- app/templates/components/table.hbs -->
<Hds::Table
  @model={{this.model.data}}
  @columns={{array
      (hash 
        key='status'
        label='Status'
        isSortable=true
        sortingFunction=this.myCustomSortingFunction
      )
      (hash key='album' label='Album')
      (hash key='year' label='Year')
    }}
  @onSort={{this.myCustomOnSort}}
>
  <!-- <:body> here -->
</Hds::Table>
```

Here’s an example of what a custom sort function could look like. In this example, we are indicating that we want to sort on a status, which takes its order based on the position in the array:

```javascript

> import Controller from '@ember/controller';
> import { action } from '@ember/object';
> import { tracked } from '@glimmer/tracking';
> 

Reviewer: we don't have to port all the controller code here, let's focus on what is relevant
```suggestion
```

> // we use an array to declare the custom sorting order for the clusters' status

Reviewer: ```suggestion
// we use an array to declare the custom sorting order for the "status" column
```

const customSortingCriteriaArray = [
  'failing',
  'active',
  'establishing',
  'pending',
];

> export default class ComponentsTableController extends Controller {
>   @tracked customSortOrderForClusterStatus = 'asc';

Reviewer: ```suggestion
// we track the sorting order, so it can be used in the custom sorting function
@tracked customSortOrderForStatus = 'asc';
```

>   get clustersWithExtraData() {
>     return this.model.clusters.map((record) => {
>       return {
>         ...record,
>         'status-sort-order': customSortingCriteriaArray.indexOf(
>           record['status']
>         ),
>       };
>     });
>   }

Reviewer: not needed
```suggestion
```

>   get customSortingMethodForClusterStatus() {
>     return (s1, s2) => {
>       const index1 = customSortingCriteriaArray.indexOf(s1['status']);
>       const index2 = customSortingCriteriaArray.indexOf(s2['status']);
>       if (index1 < index2) {
>         return this.customSortOrderForClusterStatus === 'asc' ? -1 : 1;
>       } else if (index1 > index2) {
>         return this.customSortOrderForClusterStatus === 'asc' ? 1 : -1;
>       } else {
>         return 0;
>       }
>     };
>   }

Reviewer: ```suggestion
// we define a "getter" that returns a custom sorting function ("s1" and "s2" are data records)
get customSortingMethodForStatus() {
  return (s1, s2) => {
    const index1 = customSortingCriteriaArray.indexOf(s1['status']);
    const index2 = customSortingCriteriaArray.indexOf(s2['status']);
    if (index1 < index2) {
      return this.customSortOrderForStatus === 'asc' ? -1 : 1;
    } else if (index1 > index2) {
      return this.customSortOrderForStatus === 'asc' ? 1 : -1;
    } else {
      return 0;
    }
  };
}
```

>   @action
>   customOnSort(_sortBy, sortOrder) {
>     this.customSortOrderForClusterStatus = sortOrder;
>   }
> }

Reviewer: ```suggestion
// we define a callback function that listens to the `onSort` event in the table, and updates the tracked sort order values accordingly
@action
customOnSort(_sortBy, sortOrder) {
  this.customSortOrderForStatus = sortOrder;
}
```

</Hds::Table>
```

#### Replacing components as a pre-adoption step

> If you're not quite ready to replace your existing tables with this component, you can totally try out a pre-adoption spike with just the components themselves. It's a little more typing but it should give you an idea of what will work for you.

Reviewer 2: ```suggestion
If you’re not quite ready to replace your existing tables with this component, you can totally try out a pre-adoption spike with just the components themselves. It’s a little more typing but it should give you an idea of what will work for you.
```

</Hds::Table>
```

#### Replacing components as a pre-adoption step

> If you’re not quite ready to replace your existing tables with this component, you can totally try out a pre-adoption spike with just the components themselves. It’s a little more typing but it should give you an idea of what will work for you.

Reviewer: Two more considerations:
- the example below doesn't work (`@columns` is required, when there is a `@model`)
- I don't understand this example/paragraph. What makes this example a pre-adoption step? is using the `@model` (so is looping over records) and the yielded/contextual components.
Maybe you wanted to have an example where the table is generated using the non-yielded components? something like this?
```
<Hds::Table>
  <thead>
    <Hds::Table::Tr>
      <Hds::Table::Th>Artist</Hds::Table::Th>
      <Hds::Table::Th>Album</Hds::Table::Th>
      <Hds::Table::Th>Release Year</Hds::Table::Th>
    </Hds::Table::Tr>
  </thead>
  <tbody>
    {{#each this.myData as |record|}}
      <Hds::Table::Tr>
        <Hds::Table::Td>{{record.artist}}</Hds::Table::Td>
        <Hds::Table::Td>{{record.album}}</Hds::Table::Td>
        <Hds::Table::Td>{{record.year}}</Hds::Table::Td>
      </Hds::Table::Tr>
    {{/each}}
  </tbody>
</Hds::Table>
```

Reviewer: Created a ticket for this: https://hashicorp.atlassian.net/browse/HDS-1585
